servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/DefaultLoadBalancer.java

```diff
@@ -507,10 +513,38 @@ private void sequentialUpdateUsedHosts(List<PrioritizedHostImpl<ResolvedAddress,
             host.loadBalancingWeight(host.serviceDiscoveryWeight());
         }
         nextHosts = priorityStrategy.prioritize(nextHosts);
+        nextHosts = makeSubset(nextHosts);
         this.hostSelector = hostSelector.rebuildWithHosts(nextHosts);
         loadBalancerObserver.onHostSetChanged(Collections.unmodifiableList(nextHosts));
     }
 
+    private List<PrioritizedHostImpl<ResolvedAddress, C>> makeSubset(
+            final List<PrioritizedHostImpl<ResolvedAddress, C>> nextHosts) {
+        if (nextHosts.size() <= subsetSize) {
+            return nextHosts;
+        }
+
+        // We need to sort, and then return the list with the subsetSize number of healthy elements.
+        ArrayList<PrioritizedHostImpl<ResolvedAddress, C>> result = new ArrayList<>(nextHosts);
+        result.sort((a, b) -> Long.compare(a.randomSeed, b.randomSeed));
+
+        // We don't want to consider the unhealthy elements to be a part of our subset, so we're going to grow it
+        // to account for un-health endpoints. However, we need to know how many that is.
+        for (int i = 0, healthyCount = 0; i < result.size(); i++) {
+            if (result.get(i).isHealthy()) {
```

Comment:
Second thought: if there is a temporary blip in healthiness, we may end up with a much bigger subset size when those turn healthy again. That might be unexpected for users. WDYT if instead we create a new `result` list where we first add only `ACTIVE` hosts and add `EXPIRED` or `UNHEALTY` in the tail if number of `ACTIVE` hosts was less than `subsetSize`? We can then shuffle if necessary.

Reply:
I'd prefer to leave it the way it is for `EXPIRED` for a few reasons, the primary of which is that if we're using EXPIRED its because we're using UDP DNS which is honestly going to be extremely tough to properly do CSLB with subsetting so I'd rather not get into the complexity of that.

Reply:
Agreed about `EXPIRED`. Even though it's possible that any other SD impl also uses `EXPIRED` status, we can revisit that when we have a case. 

WDYT about a blip of unhealthy hosts and unexpected subset size increase when they turn back to active?

Reply:
Good question about size increase when hosts go unhealthy. What happens now is
```
H denotes known health endpoint and will have a pool. 
U denotes unhealthy endpoint with a pool that hasn't been active closed, but won't receive traffic.
? denotes an endpoint presumed healthy but doesn't receive traffic nor has an established connection pool.
(..) is the entire set
[..] denotes the subset eligible for selection

initial state: 
([H, H, H],?, ?, ?) 
All H hosts will have a endpoint connection pool.

one endpoint goes unhealthy so we expand to accommodate which to give us:
([H, H, U, H], ?, ?)
We will establish a connection pool to the index 3 host now as well.

That unhealthy endpoint revives, but revival happens on a timer so we transiently end up with
([H, H, H, H], ?, ?)
Now we have 4 elements in our set until the next outlier detection scan. We retained the connection pool to index 2 so we already payed the price for those connections. Many circuit breakers are based on a timer, so while we _think_ index 2 may be healthy again, it may be only for a very short time and quickly flap back to unhealthy.

The outlier detect scan runs:
([H, H, H], H, ?, ?)
Index 3 is no longer elegible for selection. These connections have already been paid for, so we just let them naturally age out. If we happen to flap again before they age out then we get lucky and don't have to pay again for that connection pool since it still remains active.
```

So if we're going to expand to new endpoints to ensure we have the specified number of _healthy_ hosts, I like the current behavior. However, there is a reasonable case to make that we shouldn't consider health when picking a subset: this makes things simpler (slightly in implementation but significantly in terms of reasoning) and can also help isolate cases where a recently introduced bad client which is killing upstreams has a limited blast radius. However, this does come at the cost of some extra resiliency.

Reply:
Thanks for clarifying. Then let's leave it as-is. Hopefully "unhealthy" is a rare state for most of the time and if everything goes crazy, we will anyway resize back as soon as outlier detect scan runs. So, the increased subset size should be short lived IIUC.